src/wallet3/db_schema.cpp

```diff
@@ -0,0 +1,372 @@
+#include "db_schema.hpp"
+
+#include "output.hpp"
+#include "block.hpp"
+
+#include <common/hex.h>
+#include <cryptonote_basic/cryptonote_basic.h>
+
+#include <fmt/core.h>
+#include <iostream>
+
+namespace wallet
+{
+  WalletDB::~WalletDB()
+  {
+  }
+
+  void
+  WalletDB::create_schema(cryptonote::network_type nettype)
+  {
+    if (db.tableExists("outputs"))
+    {
+      if (auto stored_nettype = this->network_type(); stored_nettype != nettype)
+      {
+        std::string err = fmt::format("Loaded wallet on network type \"{}\" but db has network type \"{}\"",
+            cryptonote::network_type_to_string(nettype),
+            cryptonote::network_type_to_string(stored_nettype));
+        //TODO: log error as well
+        throw std::invalid_argument(err);
+      }
+      return;
+    }
+
+    SQLite::Transaction db_tx(db);
+
+    // TODO: set up removal triggers
+    // TODO: table for balance "per account"
+    db.exec(
+        R"(
+          -- CHECK (id = 0) restricts this table to a single row
+          CREATE TABLE metadata (
+            id INTEGER NOT NULL PRIMARY KEY CHECK (id = 0),
+            db_version INTEGER NOT NULL DEFAULT 0,
+            nettype TEXT NOT NULL DEFAULT "testnet",
+            balance INTEGER NOT NULL DEFAULT 0,
+            unlocked_balance INTEGER NOT NULL DEFAULT 0,
+            last_scan_height INTEGER NOT NULL DEFAULT -1,
+            scan_target_hash TEXT NOT NULL,
+            scan_target_height INTEGER NOT NULL DEFAULT 0,
+            output_count INTEGER NOT NULL DEFAULT 0
+          );
+
+          -- insert metadata row as default
+          INSERT INTO metadata VALUES (0,0,"testnet",0,0,-1,"",0,0);
+
+          CREATE TABLE blocks (
+            height INTEGER NOT NULL PRIMARY KEY,
```

Comment:
you probably want the primary key to be height and hash so you can track altchains and rewind easier.